pkg/experiment/metrics/observer.go

```diff
@@ -0,0 +1,197 @@
+package metrics
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/prompb"
+
+	metastorev1 "github.com/grafana/pyroscope/api/gen/proto/go/metastore/v1"
+	"github.com/grafana/pyroscope/pkg/experiment/block"
+	phlaremodel "github.com/grafana/pyroscope/pkg/model"
+)
+
+type SampleObserver struct {
+	exporter          *Exporter
+	tenant            string
+	Recordings        []*Recording
+	recordingTime     int64
+	pyroscopeInstance string
+	logger            log.Logger
+}
+
+type Recording struct {
+	rule  RecordingRule
+	data  map[model.Fingerprint]*Sample
+	state *recordingState
+}
+
+type recordingState struct {
+	fp      *model.Fingerprint
+	matches bool
+	sample  *Sample
+}
+
+func NewSampleObserver(meta *metastorev1.BlockMeta, logger log.Logger) *SampleObserver {
+	recordingTime := int64(ulid.MustParse(meta.Id).Time())
+	pyroscopeInstance := pyroscopeInstanceHash(meta.Shard, meta.CreatedBy)
+	return &SampleObserver{
+		recordingTime:     recordingTime,
+		pyroscopeInstance: pyroscopeInstance,
+		logger:            logger,
+		exporter:          &Exporter{},
+	}
+}
+
+func pyroscopeInstanceHash(shard uint32, createdBy int32) string {
+	buf := make([]byte, 0, 8)
+	buf = append(buf, byte(shard>>24), byte(shard>>16), byte(shard>>8), byte(shard))
+	buf = append(buf, byte(createdBy>>24), byte(createdBy>>16), byte(createdBy>>8), byte(createdBy))
+	return fmt.Sprintf("%x", xxhash.Sum64(buf))
```

Comment:
You probably want to use `binary.AppendUint32`
```go
buf := make([]byte, 0, 8)
buf = binary.BigEndian.AppendUint32(buf, shard)
buf = binary.BigEndian.AppendUint32(buf, createdBy)
```
Or `Put`:
```go
buf := make([]byte, 8)
binary.BigEndian.PutUint32(buf[0:4], shard)
binary.BigEndian.PutUint32(buf[4:8], createdBy)
```
The order (including endianness) and does not matter.